File: .net/Harvbot.Arms.Driver/HarvbotArm2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Harvbot.Arms.Driver
{
    /// <summary>
    /// Reprersents Harvbot Arm Type 2.
    /// </summary>
    public sealed class HarvbotArm2 : HarvbotArmBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="HarvbotArm1"/> class.
        /// </summary>
        /// <param name="comNum">The COM number.</param>
        public HarvbotArm2(string comNum) : base(comNum)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="HarvbotArm2"/> class.
        /// </summary>
        /// <param name="provider">The arm controller provider.</param>
        public HarvbotArm2(IHarvbotArmProvider provider) : base(provider)
        {
        }

        /// <summary>
        /// Gets the arm type.
        /// </summary>
        public override HarvbotArmTypes ArmType
        {
            get
            {
                return HarvbotArmTypes.Type2;
            }
        }

        /// <summary>
        /// Gets the claw node.
        /// </summary>
        public HarvbotArmScrewNode Claw
        {
            get
            {
                return (HarvbotArmScrewNode)this.nodes[3];
            }
        }

        /// <summary>
        /// Initializes arm nodes.
        /// </summary>
        protected override void InitializeNodes()
        {
            this.nodes = new HarvbotArmNode[]
            {
                new HarvbotArmCircleNode(HarvbotArmNodeIdentifiers.Bedplate, this),
                new HarvbotArmCircleNode(HarvbotArmNodeIdentifiers.Shoulder, this),
                new HarvbotArmCircleNode(HarvbotArmNodeIdentifiers.Elbow, this),
                new HarvbotArmScrewNode(HarvbotArmNodeIdentifiers.Claw, this),
            };
        }
    }
}

File: arduino-library/HarvBotArm/src/HarvbotArm2.cpp
#include "HarvbotArm.h"
#include "HarvbotArm2.h"
#include "HarvbotArmNode.h"
#include "HarvbotArmServoNode.h"
#include "HarvbotArmCircleNode.h"
#include "HarvbotArmScrewNode.h"
#include "HarvbotArmConstants.h"

HarvbotArm2::HarvbotArm2(int beadplatePin, int shoulderPin, int elbowPin, HarvbotArmScrewNode* claw) 
: HarvbotArm(ArmType2) {

	this->nodesCount = 4;
	
	this->nodes = new HarvbotArmNode*[4];

	this->nodes[0] = new HarvbotArmServoNode(Bedplate, beadplatePin, 90, 0, 180);
	this->nodes[1] = new HarvbotArmServoNode(Shoulder, shoulderPin, 150, 5, 180);
	this->nodes[2] = new HarvbotArmServoNode(Elbow, elbowPin, 60, 20, 180);
	this->nodes[3] = claw;

	this->goToStartPosition();
}

HarvbotArm2::~HarvbotArm2() {}

void HarvbotArm2::goToStartPosition()
{
	((HarvbotArmCircleNode*)this->nodes[0])->write(90);
	((HarvbotArmCircleNode*)this->nodes[1])->write(150);
	((HarvbotArmCircleNode*)this->nodes[2])->write(60);
	((HarvbotArmScrewNode*)this->nodes[3])->goToStartPosition();
}

HarvbotArmScrewNode* HarvbotArm2::getClaw()
{
	return (HarvbotArmScrewNode*)this->getNodeByType(Claw);
}